plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java

```diff
@@ -240,55 +253,151 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
+        MongoColumnHandle pageSinkIdColumn = getPageSinkIdColumn(columns.stream().map(MongoColumnHandle::getName).collect(toImmutableList()));
+        List<MongoColumnHandle> allColumns = ImmutableList.<MongoColumnHandle>builderWithExpectedSize(columns.size() + 1)
+                .addAll(columns).add(pageSinkIdColumn).build();
+        SchemaTableName tempTable = SchemaTableName.schemaTableName(tableMetadata.getTable().getSchemaName(), generateTemporaryTableName());
+        mongoSession.createTable(tempTable, allColumns, Optional.empty());
+        closer.register(() -> mongoSession.dropTable(tempTable));
 
         return new MongoOutputTableHandle(
                 tableMetadata.getTable(),
-                columns.stream().filter(c -> !c.isHidden()).collect(toList()));
+                handleColumns,
+                Optional.of(tempTable.getTableName()),
+                Optional.of(pageSinkIdColumn.getName()));
     }
 
     @Override
     public Optional<ConnectorOutputMetadata> finishCreateTable(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
     {
+        MongoOutputTableHandle handle = (MongoOutputTableHandle) tableHandle;
+        if (handle.getTemporaryTableName().isPresent()) {
+            finishInsert(handle.getSchemaTableName(), handle.getTemporarySchemaTableName().get(), handle.getPageSinkIdColumnName().get(), fragments);
+        }
         clearRollback();
         return Optional.empty();
     }
 
     @Override
     public ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> insertedColumns, RetryMode retryMode)
     {
-        if (retryMode != NO_RETRIES) {
-            throw new TrinoException(NOT_SUPPORTED, "This connector does not support query retries");
+        MongoTable table = mongoSession.getTable(((MongoTableHandle) tableHandle).getSchemaTableName());
+        MongoTableHandle handle = table.getTableHandle();
+        List<MongoColumnHandle> columns = table.getColumns();
+        List<MongoColumnHandle> handleColumns = columns.stream()
+                .filter(column -> !column.isHidden())
+                .peek(column -> validateColumnNameForInsert(column.getName()))
+                .collect(toImmutableList());
+
+        if (retryMode == RetryMode.NO_RETRIES) {
+            return new MongoInsertTableHandle(
+                    handle.getSchemaTableName(),
+                    handleColumns,
+                    Optional.empty(),
+                    Optional.empty());
         }
+        MongoColumnHandle pageSinkIdColumn = getPageSinkIdColumn(columns.stream().map(MongoColumnHandle::getName).collect(toImmutableList()));
+        List<MongoColumnHandle> allColumns = ImmutableList.<MongoColumnHandle>builderWithExpectedSize(columns.size() + 1)
+                .addAll(columns).add(pageSinkIdColumn).build();
```

Comment:
L288-289 is also the same.
```suggestion
        List<MongoColumnHandle> allColumns = ImmutableList.<MongoColumnHandle>builderWithExpectedSize(columns.size() + 1)
                .addAll(columns)
                .add(pageSinkIdColumn)
                .build();
```